tests/acceptance/test_sro_measures_dsl.py

```diff
@@ -0,0 +1,157 @@
+    copd_codelist,
+    hba1c_codelist,
+    medication_review_codelist,
+    qrisk_codelist,
+    rbc_codelist,
+    sodium_codelist,
+    systolic_bp_codelist,
+    tsh_codelist,
+)
+
+
+def test_sro_measures_study(cohort, subtest):
+    index_date = "2020-11-01"
+    end_of_index_month = "2020-11-30"
+
+    measures_codelists = {
+        "medication_review": medication_review_codelist,
+        "systolic_bp": systolic_bp_codelist,
+        "qrisk": qrisk_codelist,
+        "cholesterol": cholesterol_codelist,
+        "alt": alt_codelist,
+        "tsh": tsh_codelist,
+        "rbc": rbc_codelist,
+        "hba1c": hba1c_codelist,
+        "sodium": sodium_codelist,
+        "asthma": asthma_codelist,
+        "copd": copd_codelist,
+    }
+
+    with subtest("population"):
+        filtered_practice_registrations = (
+            practice_registrations.filter(
+                practice_registrations.date_start <= index_date
+            )
+            # TODO: Reinstate the following when Predicates support logic operations
+            # .filter(
+            #     (practice_registrations.date_end >= index_date)
+            #     | (practice_registrations.date_end != None)
+            # )
+            .filter(practice_registrations.date_end >= index_date)
+        )
+
+        is_registered = filtered_practice_registrations.exists_for_patient()
+
+        # TODO: The following would be nice
+        # not_dead = patient_demographics.date_of_death == None
+
+        # TODO: Does this work as we expect it to work?
+        # not_dead = (
+        #     patient_demographics.select_column(patient_demographics.date_of_death)
+        #     == None  # noqa: E711
+        # )
+
+        # TODO: This does not work, because population.value is an instance of
+        # Comparator
+        # cohort.set_population(is_registered & not_dead)
+
+        cohort.set_population(is_registered)
+
+    with subtest(
+        "practice_id", missing_feature="add practice ids to practice registrations"
+    ):
+        cohort.practice_id = (
+            filtered_practice_registrations.sort_by(practice_registrations.date_end)
+            .last_for_patient()
+            .select_column(practice_registrations.practice_id)
+        )
+        cohort.assert_results(practice_id=1)
+
+    with subtest("codelists"):
+        for measure_name, target_codelist in measures_codelists.items():
+            filtered_clinical_events = (
+                clinical_events.filter(
+                    clinical_events.code.is_in(target_codelist.codes)
+                )
+                .filter(clinical_events.date >= index_date)
+                .filter(clinical_events.date <= end_of_index_month)
+            )
+
+            cohort.add_variable(
+                measure_name, filtered_clinical_events.exists_for_patient()
+            )
+
+            # TODO: SortedEventFrame.exists_for_patient doesn't exist, so we can't
+            # filter and sort together.
```

Comment:
I don't understand this. Can we discuss? We don't think that this should be necessary.

Reply:
I thought I could do something like this:

```python
# wrangle the clinical events that interest us
events = clinical_events.filter(
    # codelist filter
).filter(
    # index date filter
).filter(
    # end of index month filter
).sort_by(
    # date sort
)

# use them to add variables to our cohort
cohort.add_variable("some_measure", events.exists_for_patient())
cohort.add_variable("some_measure_event_code", events.last_for_patient().select_column(events.code)
```

But `events` is an instance of `SortedEventFrame`, which doesn't have a `exists_for_patient` method 🙁

Reply:
Right. The reason for that is that the _only_ reason to sort an event frame is because you're then going to pick the first or last one. There is never any need to sort and then aggregate with exists or count. So you'd write this instead:

```py
events = clinical_events.filter(
    # codelist filter
).filter(
    # index date filter
).filter(
    # end of index month filter
)

cohort.add_variable("some_measure", events.exists_for_patient())
cohort.add_variable("some_measure_event_code", events.sort_by(events.date).last_for_patient().select_column(events.code)
```

Reply:
(Your version isn't semantically incorrect, but there is a redundant sort. We're trying to define the DSL in such a way that they can only ever do operations that make sense.)

Reply:
Thanks for explaining the rationale for not implementing `SortedEventFrame.exists_for_patient`. I'm not sure that "no redundancy" maps to "making sense", though. In this case, it made sense to me to wrangle the clinical events before using them to add variables to the cohort; and as I was going to sort them anyway, the sort certainly didn't feel redundant.